File: src/995.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int minKBitFlips(int* nums, int numsSize, int k) {
    int retVal = 0;

    int revCnt = 0;
    int i;
    for (i = 0; i < numsSize; ++i) {
        if ((i >= k) && (nums[i - k] > 1)) {
            revCnt ^= 1;
            nums[i - k] -= 2;
        }

        if (nums[i] == revCnt) {
            if (i + k > numsSize) {
                retVal = -1;
                break;
            }
            ++retVal;
            revCnt ^= 1;
            nums[i] += 2;
        }
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
        int k;
    } testCase[] = {{{0, 1, 0}, 3, 1}, {{1, 1, 0}, 3, 2}, {{0, 0, 0, 1, 0, 1, 1, 0}, 8, 3}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums = [0,1,0], k = 1
     *  Output: 2
     *
     *  Input: nums = [1,1,0], k = 2
     *  Output: -1
     *
     *  Input: nums = [0,0,0,1,0,1,1,0], k = 3
     *  Output: 3
     */

    int answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums[j]);
        }
        printf("], k = %d\n", testCase[i].k);

        answer = minKBitFlips(testCase[i].nums, testCase[i].numsSize, testCase[i].k);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/995.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def minKBitFlips(self, nums: List[int], k: int) -> int:
        retVal = 0

        numsSize = len(nums)
        revCnt = 0
        for i in range(numsSize):
            if (i >= k) and (nums[i - k] > 1):
                revCnt ^= 1
                nums[i - k] -= 2

            if nums[i] == revCnt:
                if i + k > numsSize:
                    retVal = -1
                    break
                retVal += 1
                revCnt ^= 1
                nums[i] += 2

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, k in zip([[0, 1, 0], [1, 1, 0], [0, 0, 0, 1, 0, 1, 1, 0]], [1, 2, 3]):
            # /* Example
            #  *  Input: nums = [0,1,0], k = 1
            #  *  Output: 2
            #  *
            #  *  Input: nums = [1,1,0], k = 2
            #  *  Output: -1
            #  *
            #  *  Input: nums = [0,0,0,1,0,1,1,0], k = 3
            #  *  Output: 3
            #  */
            logging.info("Input: nums = %s, k = %s", nums, k)

            retVal = pSolution.minKBitFlips(nums, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
